contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,345 @@
+use the Kubernetes API or kubectl to gracefully delete a Pod.
+1. As a Reliability Engineer for an application running on a Kubernetes cluster,
+I want to configure Pods to communicate the reason that they were terminated via 
+a HTTP GET request so that I can configure monitoring and alerting based on 
+historic trends.
+
+## Assumptions
+
+1. Kubelet will always invoke the pre-stop handler prior to sending a `TERM` 
+signal to the entry point of a container.
+1. pre-stop handlers will not contain complicated or long running business logic. 
+The logic associated with container cleanup will be executed by the container's
+entry point during the container's configured termination grace period.
+
+## Requirements
+
+1. Termination reason must be available to a container at the time its pre-stop 
+handler is invoked. In this way, applications can perform any actions
+necessary to configure the container's entry point prior to the entry point 
+receiving the `TERM` signal.
+1. The delivery mechanism for termination reason must be consumable by, and 
+ compatible with, existing HTTP and command based pre-stop handlers.
+1. Any headers, variables, or parameters used to communicate a termination 
+reason must not be hard coded, and the receiving container must be able to 
+override the defaults.
+
+## API Objects
+
+Termination reason is implemented by a string type.
+
+```golang
+// TerminationReason is a string type used to indicate the reason for 
+// termination to the preStop lifecycle hook of a Pod's containers.
+type TerminationReason string
+
+const (
+	// ReasonEviction is the default reason used to communicate that a Pod has 
+ 	// been terminated due to an eviction.
+	ReasonEviction TerminationReason = "Eviction"
+	// ReasonIntolerableTaint is the default reason used to communicate that a Pod 
+	// has been terminated due to a taint for which it has no declared toleration.
+	ReasonIntolerableTaint = "IntolerableTaint"
+	// ReasonDecommissioned is the default reason sent by controllers to 
+	// indicate that a Pod has been terminated due to a horizontal scaling 
+	// event. Pods receiving this reason should not expect to be rescheduled.
+	ReasonDecommissioned = "Decommissioned"
+	// ReasonUpdate is the reason sent by a controller or User to indicate that 
+	// a Pod has been terminated for the purposes of a destructive update. Pods 
+	// receiving this signal should expect to be rescheduled immediately. Note
+	// that this DOES NOT imply that scheduling will succeed.
+	ReasonUpdate = "Update"
+)
+
+```
+
+DeleteOptions is modified to carry the termination reason.
+
+```golang
+type DeleteOptions struct {
+	// Other fields omitted for brevity.
+	
+	// Reason indicates the reason for the Pod's termination. This field may 
+	// be supplied by a user or a controller.
+	Reason *TerminationReason `json:"reason,omitempty"`
+}
+```
+
+The ObjectMeta struct is modified to carry the termination reason via 
```

Comment:
So as @liggitt suggests below we could image a world were a signal is sent to controllers and then propagated to Pods' handlers when the controller performs a deletion. If we use PodStatus, we are always constrained to Podsl. If we put it in ObjectMeta we have flexibility going forward. I'm more in favor of taking @thockin's suggestion and generalizing.